File: js/target.js
/**
 * Target Motion Engine — ISRO PS-26169.
 * Implements 4 Mandatory + 3 Optional Motion Models with configuration boundary validation.
 */

import { SimulationState, prng } from './state.js';

export class TargetEngine {
  constructor() {
    this.phase = 0;
    this.randomVx = 60;
    this.randomVy = 40;
    this.sineDirection = 1;
  }

  reset() {
    this.phase = 0;
    const tgt = SimulationState.target;
    const env = SimulationState.environment;
    
    // Position according to initial location setting
    if (tgt.initialLocation === 'center') {
      tgt.worldX = env.width / 2;
      tgt.worldY = env.height / 2;
    } else if (tgt.initialLocation === 'custom') {
      tgt.worldX = Math.max(50, Math.min(env.width - 50, tgt.customX));
      tgt.worldY = Math.max(50, Math.min(env.height - 50, tgt.customY));
    } else {
      // Random initial location within safe margin
      tgt.worldX = prng.range(200, env.width - 200);
      tgt.worldY = prng.range(200, env.height - 200);
    }

    tgt.prevWorldX = tgt.worldX;
    tgt.prevWorldY = tgt.worldY;
    tgt.vx = 80;
    tgt.vy = 60;
    this.randomVx = prng.range(-80, 80);
    this.randomVy = prng.range(-80, 80);
    tgt.trajectoryHistory = [{ x: tgt.worldX, y: tgt.worldY }];
  }

  update(dt, elapsedTime) {
    const tgt = SimulationState.target;
    const env = SimulationState.environment;
    tgt.prevWorldX = tgt.worldX;
    tgt.prevWorldY = tgt.worldY;

    const centerX = env.width / 2;
    const centerY = env.height / 2 + 150; // Offset above bottom camera station
    const omega = 2 * Math.PI * tgt.frequency;

    let newX = tgt.worldX;
    let newY = tgt.worldY;

    switch (tgt.motionType) {
      // 1. Mandatory: Straight Line
      case 'straight': {
        newX += tgt.vx * dt;
        newY += tgt.vy * dt;
        // Enforce boundary reflection
        const margin = 100;
        if (newX <= margin) { newX = margin; tgt.vx = Math.abs(tgt.vx); }
        if (newX >= env.width - margin) { newX = env.width - margin; tgt.vx = -Math.abs(tgt.vx); }
        if (newY <= margin) { newY = margin; tgt.vy = Math.abs(tgt.vy); }
        if (newY >= env.height - margin) { newY = env.height - margin; tgt.vy = -Math.abs(tgt.vy); }
        break;
      }

      // 2. Mandatory: Circular
      case 'circular': {
        const r = Math.min(tgt.radius, env.width / 2 - 120);
        newX = centerX + r * Math.cos(omega * elapsedTime);
        newY = centerY + r * Math.sin(omega * elapsedTime);
        break;
      }

      // 3. Mandatory: Figure of 8 (Lemniscate parametric curve)
      case 'figure8': {
        const r = Math.min(tgt.radius, env.width / 2 - 150);
        newX = centerX + r * Math.sin(omega * elapsedTime);
        newY = centerY + (r * 0.5) * Math.sin(2 * omega * elapsedTime);
        break;
      }

      // 4. Mandatory: Random Walk with Inertia
      case 'random': {
        // Subtle acceleration drift
        this.randomVx += prng.range(-30, 30) * dt;
        this.randomVy += prng.range(-30, 30) * dt;
        const maxV = tgt.speed;
        this.randomVx = Math.max(-maxV, Math.min(maxV, this.randomVx));
        this.randomVy = Math.max(-maxV, Math.min(maxV, this.randomVy));

        newX += this.randomVx * dt;
        newY += this.randomVy * dt;

        const margin = 100;
        if (newX <= margin) { newX = margin; this.randomVx = Math.abs(this.randomVx); }
        if (newX >= env.width - margin) { newX = env.width - margin; this.randomVx = -Math.abs(this.randomVx); }
        if (newY <= margin) { newY = margin; this.randomVy = Math.abs(this.randomVy); }
        if (newY >= env.height - margin) { newY = env.height - margin; this.randomVy = -Math.abs(this.randomVy); }
        break;
      }

      // 5. Optional: Spiral
      case 'spiral': {
        const maxR = Math.min(tgt.radius, env.width / 2 - 120);
        const minR = 80;
        const modOmega = omega * 0.25;
        const currentR = minR + (maxR - minR) * (0.5 + 0.5 * Math.sin(modOmega * elapsedTime));
        newX = centerX + currentR * Math.cos(omega * elapsedTime);
        newY = centerY + currentR * Math.sin(omega * elapsedTime);
        break;
      }

      // 6. Optional: Sinusoidal
      case 'sinusoidal': {
        const span = env.width - 300;
        const period = span / Math.max(20, tgt.speed);
        const progress = (elapsedTime % (2 * period)) / period;
        let xOffset;
        if (progress <= 1) {
          xOffset = progress * span;
        } else {
          xOffset = (2 - progress) * span;
        }
        newX = 150 + xOffset;
        newY = centerY + (tgt.radius * 0.4) * Math.sin(omega * elapsedTime * 2);
        break;
      }

      // 7. Optional: User-defined X(t), Y(t)
      case 'userDefined': {
        try {
          const t = elapsedTime;
          // Evaluated safely in sandbox context
          const fnX = new Function('t', `return ${tgt.userFormulaX};`);
          const fnY = new Function('t', `return ${tgt.userFormulaY};`);
          newX = fnX(t);
          newY = fnY(t);
          if (isNaN(newX) || isNaN(newY)) throw new Error('NaN');
        } catch (e) {
          // Fallback to stable figure-8 if user formula errors
          const r = tgt.radius;
          newX = centerX + r * Math.sin(omega * elapsedTime);
          newY = centerY + (r * 0.5) * Math.sin(2 * omega * elapsedTime);
        }
        break;
      }

      default:
        newX = centerX;
        newY = centerY;
        break;
    }

    // Boundary clamping with margin
    tgt.worldX = Math.max(30, Math.min(env.width - 30, newX));
    tgt.worldY = Math.max(30, Math.min(env.height - 30, newY));

    // Append to trajectory history (max 300 samples)
    if (SimulationState.simulation.currentFrame % 2 === 0) {
      tgt.trajectoryHistory.push({ x: tgt.worldX, y: tgt.worldY });
      if (tgt.trajectoryHistory.length > 300) {
        tgt.trajectoryHistory.shift();
      }
    }
  }

  // Precomputes trajectory points for UI preview widgets
  getTrajectoryPath(motionType, steps = 120) {
    const env = SimulationState.environment;
    const tgt = SimulationState.target;
    const centerX = env.width / 2;
    const centerY = env.height / 2 + 150;
    const omega = 2 * Math.PI * tgt.frequency;
    const points = [];
    const totalTime = 1.0 / Math.max(0.01, tgt.frequency);

    for (let i = 0; i <= steps; i++) {
      const t = (i / steps) * totalTime;
      let px = centerX;
      let py = centerY;

      if (motionType === 'figure8') {
        const r = Math.min(tgt.radius, env.width / 2 - 150);
        px = centerX + r * Math.sin(omega * t);
        py = centerY + (r * 0.5) * Math.sin(2 * omega * t);
      } else if (motionType === 'circular') {
        const r = Math.min(tgt.radius, env.width / 2 - 120);
        px = centerX + r * Math.cos(omega * t);
        py = centerY + r * Math.sin(omega * t);
      } else if (motionType === 'straight') {
        px = 200 + (env.width - 400) * (i / steps);
        py = centerY - 150 + (300) * (i / steps);
      } else if (motionType === 'spiral') {
        const currentR = 80 + (tgt.radius - 80) * (i / steps);
        px = centerX + currentR * Math.cos(omega * t * 3);
        py = centerY + currentR * Math.sin(omega * t * 3);
      } else if (motionType === 'sinusoidal') {
        px = 150 + (env.width - 300) * (i / steps);
        py = centerY + (tgt.radius * 0.4) * Math.sin(omega * t * 3);
      } else if (motionType === 'random') {
        // Decorative representation of random walk
        px = centerX + Math.sin(i * 0.15) * 200 + Math.cos(i * 0.4) * 80;
        py = centerY + Math.cos(i * 0.2) * 160 + Math.sin(i * 0.3) * 60;
      } else if (motionType === 'userDefined') {
        try {
          const fnX = new Function('t', `return ${tgt.userFormulaX};`);
          const fnY = new Function('t', `return ${tgt.userFormulaY};`);
          px = fnX(t);
          py = fnY(t);
        } catch (e) {
          px = centerX; py = centerY;
        }
      }
      points.push({ x: px, y: py });
    }
    return points;
  }
}

File: backend/target.py
"""
Target Generator & Trajectory Engine — ISRO PS-26169 Backend.
Implements the 4 mandatory + 3 optional motion models, spot intensity rendering
(5-20 px, Square default, Circle, Rectangle), and ground-truth telemetry generation.
"""

from __future__ import annotations

import dataclasses
import math
from typing import Dict, List, Literal, Tuple
import numpy as np

from backend.environment import VirtualEnvironment

MotionType = Literal["straight", "circular", "figure8", "random", "spiral", "sinusoidal", "user_defined"]
ShapeType = Literal["Square", "Circle", "Rectangle"]


@dataclasses.dataclass
class TargetState:
    x: float
    y: float
    vx: float
    vy: float
    shape: ShapeType
    size: int  # 5 to 20 px


class TargetGenerator:
    """Simulates moving optical beacon target inside the virtual environment."""

    def __init__(
        self,
        env: VirtualEnvironment,
        target_id: str = "T1",
        shape: ShapeType = "Square",
        size: int = 10,
        initial_location: str = "random",
        custom_pos: Tuple[float, float] | None = None,
        motion_type: MotionType = "figure8",
        speed: float = 120.0,
        radius: float = 450.0,
        frequency: float = 0.12,
        seed: int = 26169,
    ) -> None:
        self.env = env
        self.target_id = target_id
        self.shape: ShapeType = shape
        self.size = max(5, min(20, size))  # Bound strictly to [5, 20] px
        self.initial_location = initial_location
        self.custom_pos = custom_pos or (1000.0, 1000.0)
        self.motion_type: MotionType = motion_type
        self.speed = speed
        self.radius = radius
        self.frequency = frequency
        self.seed = seed

        self.rng = np.random.default_rng(self.seed)

        # Runtime dynamic state
        self.x = 1000.0
        self.y = 1000.0
        self.vx = 80.0
        self.vy = 60.0
        self.random_vx = 60.0
        self.random_vy = 40.0
        self.history: List[Tuple[float, float]] = []

        self.reset()

    def reset(self) -> None:
        """Resets target position and dynamics to initial state."""
        self.rng = np.random.default_rng(self.seed)

        if self.initial_location == "center":
            self.x = self.env.width / 2.0
            self.y = self.env.height / 2.0
        elif self.initial_location == "custom":
            self.x, self.y = self.env.clamp_bounds(self.custom_pos[0], self.custom_pos[1], margin=50.0)
        else:  # "random"
            self.x = float(self.rng.uniform(200, self.env.width - 200))
            self.y = float(self.rng.uniform(200, self.env.height - 200))

        self.vx = 80.0
        self.vy = 60.0
        self.random_vx = float(self.rng.uniform(-80, 80))
        self.random_vy = float(self.rng.uniform(-80, 80))
        self.history = [(self.x, self.y)]

    def update(self, dt: float, elapsed_time: float) -> TargetState:
        """Advances target position along the selected trajectory model."""
        center_x = self.env.width / 2.0
        center_y = self.env.height / 2.0 + 150.0  # Centered in upper workspace
        omega = 2.0 * math.pi * self.frequency

        new_x = self.x
        new_y = self.y

        # 1. Mandatory: Straight Line
        if self.motion_type == "straight":
            new_x += self.vx * dt
            new_y += self.vy * dt
            margin = 100.0
            if new_x <= margin:
                new_x = margin
                self.vx = abs(self.vx)
            elif new_x >= self.env.width - margin:
                new_x = self.env.width - margin
                self.vx = -abs(self.vx)
            if new_y <= margin:
                new_y = margin
                self.vy = abs(self.vy)
            elif new_y >= self.env.height - margin:
                new_y = self.env.height - margin
                self.vy = -abs(self.vy)

        # 2. Mandatory: Circular
        elif self.motion_type == "circular":
            r = min(self.radius, self.env.width / 2.0 - 120.0)
            new_x = center_x + r * math.cos(omega * elapsed_time)
            new_y = center_y + r * math.sin(omega * elapsed_time)

        # 3. Mandatory: Figure of 8 (Lemniscate)
        elif self.motion_type == "figure8":
            r = min(self.radius, self.env.width / 2.0 - 150.0)
            new_x = center_x + r * math.sin(omega * elapsed_time)
            new_y = center_y + (r * 0.5) * math.sin(2.0 * omega * elapsed_time)

        # 4. Mandatory: Random Walk with Inertia
        elif self.motion_type == "random":
            self.random_vx += float(self.rng.uniform(-30, 30)) * dt
            self.random_vy += float(self.rng.uniform(-30, 30)) * dt
            max_v = self.speed
            self.random_vx = max(-max_v, min(max_v, self.random_vx))
            self.random_vy = max(-max_v, min(max_v, self.random_vy))

            new_x += self.random_vx * dt
            new_y += self.random_vy * dt

            margin = 100.0
            if new_x <= margin:
                new_x = margin
                self.random_vx = abs(self.random_vx)
            elif new_x >= self.env.width - margin:
                new_x = self.env.width - margin
                self.random_vx = -abs(self.random_vx)
            if new_y <= margin:
                new_y = margin
                self.random_vy = abs(self.random_vy)
            elif new_y >= self.env.height - margin:
                new_y = self.env.height - margin
                self.random_vy = -abs(self.random_vy)

        # 5. Optional: Spiral
        elif self.motion_type == "spiral":
            max_r = min(self.radius, self.env.width / 2.0 - 120.0)
            min_r = 80.0
            mod_omega = omega * 0.25
            current_r = min_r + (max_r - min_r) * (0.5 + 0.5 * math.sin(mod_omega * elapsed_time))
            new_x = center_x + current_r * math.cos(omega * elapsed_time)
            new_y = center_y + current_r * math.sin(omega * elapsed_time)

        # 6. Optional: Sinusoidal
        elif self.motion_type == "sinusoidal":
            span = self.env.width - 300.0
            period = span / max(20.0, self.speed)
            progress = (elapsed_time % (2.0 * period)) / period
            x_offset = progress * span if progress <= 1.0 else (2.0 - progress) * span
            new_x = 150.0 + x_offset
            new_y = center_y + (self.radius * 0.4) * math.sin(omega * elapsed_time * 2.0)

        # 7. Optional: User-defined
        elif self.motion_type == "user_defined":
            t = elapsed_time
            # Default safe formula
            new_x = center_x + 400.0 * math.sin(t * 0.4)
            new_y = center_y + 300.0 * math.sin(t * 0.8)

        # Enforce boundary clamping
        self.x, self.y = self.env.clamp_bounds(new_x, new_y, margin=30.0)
        self.history.append((self.x, self.y))
        if len(self.history) > 500:
            self.history.pop(0)

        return TargetState(
            x=self.x,
            y=self.y,
            vx=self.vx if self.motion_type == "straight" else (self.random_vx if self.motion_type == "random" else 0.0),
            vy=self.vy if self.motion_type == "straight" else (self.random_vy if self.motion_type == "random" else 0.0),
            shape=self.shape,
            size=self.size,
        )

    def render_spot_matrix(self) -> np.ndarray:
        """
        Renders a 2D optical intensity matrix of the beacon spot.
        Returns a float32 array with shape (size * 3, size * 3) centered at the peak.
        """
        sz = self.size
        dim = sz * 3
        center = dim / 2.0 - 0.5
        grid_y, grid_x = np.ogrid[:dim, :dim]

        dist_sq = (grid_x - center) ** 2 + (grid_y - center) ** 2

        # 1. Optical Bloom (Gaussian optical diffraction falloff)
        sigma = sz * 0.6
        intensity = 255.0 * np.exp(-dist_sq / (2.0 * sigma**2))

        # 2. Shape Footprint Core
        if self.shape == "Circle":
            core_mask = dist_sq <= (sz / 2.0) ** 2
            intensity[core_mask] = 255.0
        elif self.shape == "Rectangle":
            core_mask = (abs(grid_x - center) <= sz / 2.0) & (abs(grid_y - center) <= sz / 3.0)
            intensity[core_mask] = 255.0
        else:  # Square (Default)
            core_mask = (abs(grid_x - center) <= sz / 2.0) & (abs(grid_y - center) <= sz / 2.0)
            intensity[core_mask] = 255.0

        return np.clip(intensity, 0.0, 255.0).astype(np.float32)
